More robust checking for timeout exceptions

dpsutton opened the pull request:
Overview:
We enumerate items in the database and then we want to see which ones
the user can actually run queries against. To do this we run a query
like `select false from TABLE where 1 != 1 limit 0`. The hope is that
the query analyzer realizes it only needs to do permission checks and
doesn't try to satisfy these conditions (either because limit says no
rows, or 1 != 1 is easily proven to filter all rows). So we _usually_
get very quick feedback if we can select from the table or not.

Snowflake seems to not do this on views. If a view takes a long time to
query, snowflake seems to take a long time to query. The original fix
was to add a 15 second timeout to the query. The expectation here is
that Snowflake can quickly fail if there are no permissions for the
table or view, but might take a long time to run the probe query. But we
can interpret a timeout as a successful permission check.

The original way we checked for timeouts was a very naive check for
`java.sql.SQLTimeoutException`. It seems basically no one uses this when
you do:

```clojure
(with-open [stmt (...)]
  (doto stmt
    (.setQueryTimeout 15)
    (.execute)))
```

For instance, here's some logs from Snowflake:

```
WARN net.snowflake.client.core.StmtUtil Cancelling query c6759f9f-37d0-4f71-9458-e74f2c795215 with reason TIMEOUT
INFO metabase.driver.sql-jdbc.sync.describe-database <schema>:<table-name>: Assuming no SELECT privileges: caught exception,net.snowflake.client.jdbc.SnowflakeSQLException: SQL execution canceled,	at ...
```

Note It's a regular `net.snowflake.client.jdbc.SnowflakeSQLException` but with message "SQL execution canceled".
Here's a true permissions failure for comparison:

```
sync.describe-database <schema>:<table>: Assuming no SELECT privileges: caught exception,net.snowflake.client.jdbc.SnowflakeSQLException: SQL compilation error:,Failure during expansion of view 'SOME_VIEW': SQL compilation error:,Object 'DB.SCHEMA.TABLE' does not exist or not authorized.,
```

So we use some regex checks against the messages to try to detect when
we have a permissions failure (cannot select) vs a timeout of the
probe (presumably can select).

And a reminder of the balances here: it's better to show someone a table
that they get an error if they try to query than to hide a table that
they can query. So this change should bias towards reporting can-select
in cases where it might not be clear.

dpsutton committed:
wip

dpsutton committed:
More robust checking for timeout exceptions

Overview:
We enumerate items in the database and then we want to see which ones
the user can actually run queries against. To do this we run a query
like `select false from TABLE where 1 != 1 limit 0`. The hope is that
the query analyzer realizes it only needs to do permission checks and
doesn't try to satisfy these conditions (either because limit says no
rows, or 1 != 1 is easily proven to filter all rows). So we _usually_
get very quick feedback if we can select from the table or not.

Snowflake seems to not do this on views. If a view takes a long time to
query, snowflake seems to take a long time to query. The original fix
was to add a 15 second timeout to the query. The expectation here is
that Snowflake can quickly fail if there are no permissions for the
table or view, but might take a long time to run the probe query. But we
can interpret a timeout as a successful permission check.

The original way we checked for timeouts was a very naive check for
`java.sql.SQLTimeoutException`. It seems basically no one uses this when
you do:

```clojure
(with-open [stmt (...)]
  (doto stmt
    (.setQueryTimeout 15)
    (.execute)))
```

For instance, here's some logs from Snowflake:

```
WARN net.snowflake.client.core.StmtUtil Cancelling query 0e8987ee-2d22-40f9-b30d-0129d38ce976 with reason TIMEOUT
INFO metabase.driver.sql-jdbc.sync.describe-database <schema>:<table-name>: Assuming no SELECT privileges: caught exception,net.snowflake.client.jdbc.SnowflakeSQLException: SQL execution canceled,	at ...
```

Note It's a regular `net.snowflake.client.jdbc.SnowflakeSQLException`.
Here's a true permissions failure for comparison:

```
sync.describe-database <schema>:<table>: Assuming no SELECT privileges: caught exception,net.snowflake.client.jdbc.SnowflakeSQLException: SQL compilation error:,Failure during expansion of view 'SOME_VIEW': SQL compilation error:,Object 'DB.SCHEMA.TABLE' does not exist or not authorized.,
```

So we use some regex checks against the messages to try to detect when
we have a permissions failure (cannot select) vs a timeout of the
probe (presumably can select).

And a reminder of the balances here: it's better to show someone a table
that they get an error if they try to query than to hide a table that
they can query. So this change should bias towards reporting can-select
in cases where it might not be clear.

dpsutton committed:
Only run tests when classes are present

if you don't have :drivers on classpath it would fail

dpsutton committed:
ensure boolean is returned

tests assert false? on a nil

dpsutton committed:
CI fails based on the indentation of comments

dpsutton committed:
empty commit for CI

dpsutton committed:
Move postgres test into postgres test namespace

dpsutton commented:
Things i'm interested in feedback about (of course _anything_ but things i specifically want to call out):
- `when-class` macro lets me colocate tests even if the drivers are absent. Do you like this? I can spread it out, but it gets, well, spread out
- i moved the `:postgres` test into the postgres test. The linter didn't like hardcoding a driver here. Should i override the linter and put all the tests together in one spot? Also, I don't know how to cause a long running query for snowflake or redshift. Basically jsut for postgres with `pg_sleep`. The cloud dbs don't like you to be able to choke execution. I could possibly engineer a long query but it's just causing lots of computation.
- have a better way to accomplish this? I'm checking messages. I don't know of a better way

ericnormand commented:
> And a reminder of the balances here: it's better to show someone a table
that they get an error if they try to query than to hide a table that
they can query. So this change should bias towards reporting can-select
in cases where it might not be clear.

If this is the case, would it make more sense to test the opposite? That is, test that the message is a known permission error message instead of testing that it is a known timeout error message? Missing one of the DB's timeout messages (because they change it) would make can-select be false.

ericnormand commented:
> * have a better way to accomplish this? I'm checking messages. I don't know of a better way

From ChatGPT:

> Most real-world JDBC drivers just throw a plain SQLException, but the `getSQLState()` method on the exception gives you the SQLSTATE error code, which is more informative.
>
> For example:

* 28000 — Invalid authorization specification (bad username/password).
* 42000 — Syntax error or access rule violation.
* 42501 — Insufficient privilege (Postgres-specific, but others use similar codes).

Perhaps we could check `sqlException.getSQLState()` for either 28000 or 42xxx range? [Reference](https://www.postgresql.org/docs/current/errcodes-appendix.html).

ericnormand commented:
> Also, I don't know how to cause a long running query for snowflake or redshift. Basically jsut for postgres with `pg_sleep`. The cloud dbs don't like you to be able to choke execution. I could possibly engineer a long query but it's just causing lots of computation.

## Snowflake

```
SELECT SYSTEM$WAIT(30);
```

https://docs.snowflake.com/en/sql-reference/functions/system_wait

## Redshift

GPT says do a large self-join. Not very reliable, but it may work.

dpsutton commented on src/metabase/driver/sql_jdbc/sync/describe_database.clj:
```diff
@@ -73,6 +73,40 @@
+    (or (instance? java.sql.SQLTimeoutException e)
+        (boolean (some (fn [re] (re-find re msg))
+                       ;; Common timeout-related messages across different DBs
+                       [#"(?i)timed?\s*(?:out|exceeded)"
```
Agree. I wish people used the nice `java.sql.SQLTimeoutException` or their own wrapper around this. I'm surprised that it's not prevalent.

edpaget committed:
Replace ex-msg matching with sqlerrorcode/sqlstate

Use jdbc error codes and sql state matching to figure out if a
SQLException is caused by user action canceling the request (essentially
how setQueryTimeout works). This adds a multimethod dispatching off the
driver keyword to match up either the SQLState code or the SQLErrorCode
with the appropriate value for a given database driver. Some of these
were found by looking how QueryTimeoutException is implemented in
Hibernate and others were found from JDBC driver source code or documentation.

edpaget commented:
@dpsutton @ericnormand I kept the timeout-based testing around for this but updated the timeout checking to use sqlstate/errorcode (because different DBs seem to use them differently for this case) to look for 'query cancelled by user' states/errors since that is what the `.setQueryTimeout` on the prepared statement should realize. 

Not sure if the multimethod impls for the error code checks should live in drivers or not, but I wanted to put the main function in a non-driver namespace, since I can also reuse this check to make the cluster lock timeout handling (on the appdb) better.

edpaget committed:
fixup! handle multiple codes for mariadb

edpaget committed:
fixup! add partial query timeout test

edpaget committed:
fixup! split apart sql timeout tests

edpaget committed:
fixup! use correct sqlstate for snowflake timeout

edpaget committed:
Merge branch 'master' into handle-timeouts-in-select-priviliges

edpaget commented:
I spent some cycles looking to if 'authorization failed' sqlstates/errorcodes might be more reliable, but they turn out to also be totally different across all jdbc drivers 😭. Given that the main target here is snowflake and the timeout conceptually seems to work, I think it's worth sticking with that solution for now.

edpaget committed:
fixup! remove split out tests

edpaget committed:
fixup! remove when-class

edpaget committed:
fixup! snowflake sql state

edpaget committed:
fixup! move query-canceled? into driver code

Removes the jdbc-exceptions namespace and instead makes query-canceled a
driver multimethod that extracts and dispatches on the SQLException
class within the sql-jdbc driver.

Non-jdbc drivers can implement this method, but it is only used for
handling errors in the cluster lock using the appdb and for the sql-jdbc
describe-database sync task.

Co-Authored-By: Nathan Voxland <nathan@metabase.com>

edpaget committed:
fixup! update driver changelog add default method

edpaget committed:
fixup! merge mysql/mariadb methods

edpaget committed:
fixup! add added metadata

snoe commented on test/metabase/driver/mysql_test.clj:
```diff
@@ -815,3 +819,18 @@
                         (map #(get {"TIMESTAMP" "DATETIME"} % %)))
                    (->> binned-query   qp/process-query mt/cols (map :database_type))))))))))
+
+(deftest have-select-privelege?-timeout-test
```
Can this be `:parallel` ?

edpaget replied:
These are using a `with-redefs` to modify the generated select statement so i don't think they can be parallel.

snoe commented on test/metabase/driver/postgres_test.clj:
```diff
@@ -1604,6 +1606,22 @@
+               (try (doto stmt
+                      (.setQueryTimeout 1)
+                      (.execute))
+                    (is false "Query successfully executed. Should sleep for 5s with a timeout of 1s")
```
This is a bit confusing, I think I'd prefer a `throw`.

edpaget committed:
fixup! make test throw instead of is false
